ARROW-13561: [C++] Implement week kernel that accepts WeekOptions

rok opened the pull request:
This is to resolve [ARROW-13561](https://issues.apache.org/jira/browse/ARROW-13561`).

rok commented:
Shall we also change `DayOfWeekOptions` to `WeekOptions`? It would be a better fit overall if this change is done.

lidavidm reviewed (commented):
So essentially, ISOWeek starts from 1, Week starts from 0, and neither use DayOfWeekOptions.week_start? In that case I would just make them two optionless kernels.

rok commented:
> So essentially, ISOWeek starts from 1, Week starts from 0, and neither use DayOfWeekOptions.week_start? In that case I would just make them two optionless kernels.

To get [parity with MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_week) `DayOfWeekOptions.week_start` should be settable either 1 or 7.

lidavidm commented:
> > So essentially, ISOWeek starts from 1, Week starts from 0, and neither use DayOfWeekOptions.week_start? In that case I would just make them two optionless kernels.
> 
> To get [parity with MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_week) `DayOfWeekOptions.week_start` should be settable either 1 or 7.

Ah, ok. But unless I'm mistaken, week_start is not actually used by the week kernel.

rok commented:
> Ah, ok. But unless I'm mistaken, week_start is not actually used by the week kernel.

I'm just trying to understand if it does :)

rok commented:
This evaluates to true in MySQL on sqlfiddle:
```SQL
SELECT WEEK('2008-1-6', 1) != WEEK('2008-1-6', 4)
```

Mode | First day of week | Range | Week 1 is the first week
-- | -- | -- | --
0 | Sunday | 0-53 | with a Sunday in this year
1 | Monday | 0-53 | with 4 or more days this year
2 | Sunday | 1-53 | with a Sunday in this year
3 | Monday | 1-53 | with 4 or more days this year
4 | Sunday | 0-53 | with 4 or more days this year
5 | Monday | 0-53 | with a Monday in this year
6 | Sunday | 1-53 | with 4 or more days this year
7 | Monday | 1-53 | with a Monday in this year

lidavidm commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -337,6 +339,41 @@ struct ISOWeek {
+    } else {
+      start_week_ = mon;
+      mid_week_ = thu;
+    }
```
Do we intend to only support these two cases? If so we should validate it somewhere and make sure it's described in the docstring.

lidavidm commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -337,6 +339,41 @@ struct ISOWeek {
   Localizer localizer_;
 };
```
Given ISOWeek is a special case of Week now, maybe we should just have one kernel (and perhaps some conveniences, e.g. DayOfWeekOptions::IsoWeek())?

rok replied:
Oh yeah, that's a nice idea.
I'm still trying to exactly match `WEEK` from MySQL and once I do I'll look into this. Sorry for the WIP.

lidavidm replied:
Oh, no worries. I can hold off looking at this until you're ready.

rok replied:
Early review is great though! :)

lidavidm replied:
Ping on this? Since it seems the two kernels are very similar.

rok commented:
@ianmcook This now covers MySQL WEEK modes 1, 3, 4 and 6 (week 1 is the first week with 4 or more days this year).
If we want to cover modes 0, 2,  5, and 7 (week 1 is the first week with a Monday/Sunday in this year) we could to add another parameter e.g.: `DayOfWeekOptions.week_start_in_year`. Do you think we should?

Another thing to note here is that `DayOfWeekOptions.one_based_numbering` doesn't exactly mean we count from 0 or 1 but rather that a date from week 52 of previous year that is in this year will be counted as 0 instead of 52. It's a MySQL convention.

rok commented:
> we could to add another parameter e.g.: `DayOfWeekOptions.week_start_in_year`. 

Perhaps `DayOfWeekOptions.count_full_weeks` is more idiomatic.

lidavidm commented:
If the option doesn't apply to DayOfWeek, maybe it should be a separate WeekOptions in that case.

rok commented:
So then we now have:
`DayOfWeekOptions` with `week_start` and `one_based_numbering`

We would need to add three binary parameters to match the 8 modes of MySQL:
* `WeekOptions` with `week_start` (first day of week),`count_from_zero` (range), `count_full_weeks` (week 1 is ..)

Mode | First day of week | Range | Week 1 is the first week
-- | -- | -- | --
0 | Sunday | 0-53 | with a Sunday in this year
1 | Monday | 0-53 | with 4 or more days this year
2 | Sunday | 1-53 | with a Sunday in this year
3 | Monday | 1-53 | with 4 or more days this year
4 | Sunday | 0-53 | with 4 or more days this year
5 | Monday | 0-53 | with a Monday in this year
6 | Sunday | 1-53 | with 4 or more days this year
7 | Monday | 1-53 | with a Monday in this year

rok commented:
I tried adding `WeekOptions` with the following parameters:
```
  /// What day does the week start with (Monday=true, Sunday=false)
  bool week_starts_monday;
  /// Days in current year that fall into last years ISO week return week 0 if true
  bool count_from_zero;
  /// Is the first week fully in the the year or only its 4 or more days
  bool first_week_in_year;
```

I've also added tests with all the MySQL modes. Docs need some more work.
Shall we go this way?

ianmcook commented:
Thanks for doing this @rok! This is impressively thorough 🤩

Re the third parameter name: how about `first_week_is_fully_in_year`?

ianmcook commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -272,12 +272,27 @@ struct ARROW_EXPORT DayOfWeekOptions : public FunctionOptions {
   static DayOfWeekOptions Defaults() { return DayOfWeekOptions{}; }
 
-  /// Number days from 1 if true and from 0 if false
+  /// Number from 1 if true and from 0 if false
```
I think it's best to put back `days` in the `DayOfWeekOptions` docs since we're now using a different options class for this
```suggestion
  /// Number days from 1 if true and from 0 if false
```

rok commented:
Thanks for the fast feedback @ianmcook @lidavidm! :)

> Re the third parameter name: how about `first_week_is_fully_in_year` or `first_week_fully_in_january`?

> The third parameter's naming is a little confusing - maybe 'first_week_is_full_week' or something?

`first_week_in_year` sounds most idiomatic to me. Changed.


> For the first parameter, would it be a lot of work to allow any day of the week to be the start? It seems odd to have this be a boolean.

It wouldn't be much work no. My thinking here is: everyone starts the week with Monday or Sunday let's not even allow for bad choices (we should probably do the same for `DayOfWeekOption`). If we do know of calendars where week is starts with some other day of the week then let's introduce an integer parameter.


> The second parameter's doc is a little confusing, what is the behavior if false?

It's now changed to say:
```
  /// Dates from current year that fall into last ISO week of the previous year return
  /// 0 if true and 52 or 53 if false.
```
Does that make it clearer?

lidavidm commented:
> > For the first parameter, would it be a lot of work to allow any day of the week to be the start? It seems odd to have this be a boolean.
> 
> It wouldn't be much work no. My thinking here is: everyone starts the week with Monday or Sunday let's not even allow for bad choices (we should probably do the same for `DayOfWeekOption`). If we do know of calendars where week is starts with some other day of the week then let's introduce an integer parameter.

Ah yes, good point.

> 
> > The second parameter's doc is a little confusing, what is the behavior if false?
> 
> It's now changed to say:
> 
> ```
>   /// Dates from current year that fall into last ISO week of the previous year return
>   /// 0 if true and 52 or 53 if false.
> ```
> 
> Does that make it clearer?

Yes, thanks!

ianmcook commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -278,6 +278,22 @@ struct ARROW_EXPORT DayOfWeekOptions : public FunctionOptions {
+  /// Dates from current year that fall into last ISO week of the previous year return
+  /// 0 if true and 52 or 53 if false.
+  bool count_from_zero;
+  /// Is the first week fully in the the year or only its 4 or more days
```
```suggestion
  /// Must the first week be fully in January (true), or is a week that begins on
  /// December 29, 30, or 31 considered to be the first week of the new year (false)?
```

rok commented:
I'd like to streamline the logic and update `compute.rst` docs. I'll try to do it tomorrow morning.

lidavidm reviewed (approved):
Looks good to me, thanks.

Do we want to make ISOWeek a special case of Week? Either by having the ISOWeek kernel just call the Week kernel with the right options, or removing ISOWeek and adding a WeekOptions::IsoDefaults() or something.

rok commented:
@lidavidm yeah completely agreed.
I'm trying to make all 8 modes of Week into template parameters so that they are as efficient as ISOWeek is at the moment. Sorry it's taking a while will push asap.

pitrou commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -278,6 +278,23 @@ struct ARROW_EXPORT DayOfWeekOptions : public FunctionOptions {
+  static WeekOptions Defaults() { return WeekOptions{}; }
+
+  /// What day does the week start with (Monday=true, Sunday=false)
+  bool week_starts_monday;
```
Hmm... did you deliberately choose a different option name and semantics compared with `DayOfWeekOptions::week_start`?

rok replied:
Week start is an int and can have values from 1 to 7. Here I'd like to limit to two options - Monday/Sunday hence the boolean and different name.

pitrou replied:
Why would you like to limit it? What is the rationale for that?

rok replied:
I've not found an example of Week function that offers options other than Monday or Sunday. So I'd not like to offer more options if possible.

pitrou replied:
I'm a bit curious why it would be useful for DayOfWeek but not for Week. Ideally we should try to use similar idioms when similar concepts are involved.

rok replied:
I would actually switch `DayOfWeekOptions::week_start` to `bool DayOfWeekOptions::week_starts_monday` if everyone's up for it. Then we could just use `WeekOptions` for both and it would be just slightly wrong.

lidavidm replied:
I think that's reasonable.

pitrou commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -278,6 +278,23 @@ struct ARROW_EXPORT DayOfWeekOptions : public FunctionOptions {
+  bool week_starts_monday;
+  /// Dates from current year that fall into last ISO week of the previous year return
+  /// 0 if true and 52 or 53 if false.
+  bool count_from_zero;
```
Similar question here vs. `DayOfWeekOptions::one_based_numbering`.

rok replied:
There's a difference between the two - `one_based_numbering` controls if Monday=1 or Monday=0.
`count_from_zero` defines what is the last week of previous year that is still in this year - 0 or 52.

pitrou replied:
Er... then what is the difference between `count_from_zero` and `first_week_is_fully_in_year`?

rok replied:
`first_week_is_fully_in_year` controls if the whole week (from Sunday or Monday) is in new year or only most of it (4 days).

rok replied:
See here for all the possibilities: https://github.com/apache/arrow/pull/11026#issuecomment-910649581

rok replied:
That table was taken [from MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_week)

pitrou replied:
But why do we have to support all this? We are not trying to reimplement MySQL. And I don't see any such flexibility in PostgreSQL or SQL Server.

rok replied:
Well lubridate supports a subset of this (https://lubridate.tidyverse.org/reference/week.html). Pandas only really supports iso_week.
Perhaps @ianmcook can tweigh in on this.

pitrou replied:
"week() returns the number of complete seven day periods that have occurred between the date and January 1st, plus one" : that doesn't sound like what you're implementing, does it?

rok replied:
It could mean mode 7 (if you read complete as all days in the week starting with Monday).

rok replied:
You're right though ([source](https://rdrr.io/cran/lubridate/src/R/accessors-week.r):
```
week <- function(x)
  (yday(x) - 1) %/% 7 + 1

#' @rdname week
#' @export
"week<-" <- function(x, value)
  x <- x + days((value - week(x)) * 7)

.other_week <- function(x, week_start) {
  x <- as.POSIXlt(x)
  date <- make_date(year(x), month(x), day(x))
  wday <- wday(x, week_start = week_start)
  date <- date + (4 - wday)
  jan1 <- as.numeric(make_date(year(date), 1, 1))
  1L + (as.numeric(date) - jan1) %/% 7L
}

#' @description `isoweek()` returns the week as it would appear in the ISO 8601
#'   system, which uses a reoccurring leap week.
#' @rdname week
#' @export
isoweek <- function(x) {
  .other_week(x, 1)
}

#' @description `epiweek()` is the US CDC version of epidemiological week. It
#'   follows same rules as `isoweek()` but starts on Sunday. In other parts of
#'   the world the convention is to start epidemiological weeks on Monday,
#'   which is the same as `isoweek`.
#'
#' @rdname week
#' @export
epiweek <- function(x) {
  .other_week(x, 7)
}
```

pitrou replied:
Ha. Well, that sounds like a useful function :-)

rok replied:
It's great.

Let's try to settle on functionality we need here then.

There's the 8 modes of MySQL (they include the ISO week and epiweek) and yday/7. Which of these 9 do we want/need?

pitrou replied:
SQL server allows you to choose the starting day of the week (also, "January 1 of any year defines the starting number for the week datepart"):
https://docs.microsoft.com/en-us/sql/t-sql/functions/datepart-transact-sql?view=sql-server-ver15#week-and-weekday-datepart-arguments

PostgreSQL only seems to support ISO week numbers: https://www.postgresql.org/docs/current/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT

[Arrow](https://arrow.readthedocs.io/en/latest/) (the other one) does not seem to support non-ISO week numbers.

pitrou replied:
Note that lubridate's "epiweek" is simply an [American week number](https://savvytime.com/week-number/united-states/2021) as far as I understand.

So we could have a "week" function with configurable week start day, or more simply a "us_week" function with weeks starting on Sunday (both with semantics otherwise similar to "iso_week")

pitrou replied:
Or of course we have a "week" function allowing to configure the starting day of the week. Week number 1 is then the week containing the day `start of week + 3` (e.g. Thursday for weeks starting on Monday, Wednesday for weeks starting on Sunday...).

ianmcook replied:
My understanding is that supporting all the MySQL modes is fairly straightforward and provides the greatest flexibility. Using these three bool options to specify the modes seems like a clever way to do it.

I do not think it is important to provide a mode here that computes `day_of_year() / 7`. If someone wants that they can do `day_of_year() / 7`. We can add a binding for the lubridate `week()` function that does that.

pitrou replied:
While the implementation is reasonably simple, my main concern here is a head-scratching API. Supporting the MySQL modes does not sound like an interesting goal in itself, and exposing obscure parameters makes the API more difficult to learn and use.

We should look for actual cases. One actual use case is US week numbering (like `epiweek` in lubridate). Are there any others?

rok replied:
From [MySQL docs](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_week) about why the 8 modes:

> If a date falls in the last week of the previous year, MySQL returns 0 if you do not use 2, 3, 6, or 7 as the optional mode argument.

> One might argue that WEEK() should return 52 because the given date actually occurs in the 52nd week of 1999. WEEK() returns 0 instead so that the return value is “the week number in the given year.” This makes use of the WEEK() function reliable when combined with other functions that extract a date part from a date.

> If you prefer a result evaluated with respect to the year that contains the first day of the week for the given date, use 0, 2, 5, or 7 as the optional mode argument.

rok replied:
ping

lidavidm replied:
If the concern is primarily a confusing API, could we offer default options for 'common' modes like us_week and iso_week? Then the individual options could be left up to people who need/understand what they want.

nealrichardson commented:
Is this ready to merge?

lidavidm commented:
I'm OK with the impl here, but are we ok with the options?

jorisvandenbossche commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -942,6 +959,21 @@ Result<Datum> ISOYear(const Datum& values, ExecContext* ctx = NULLPTR);
+/// \brief Week returns week of year number for each element of `values`.
+/// First ISO week has the majority (4 or more) of its days in January.
+/// Year can have 52 or 53 weeks. Week numbering can start with 0 or 1
+/// depending on DayOfWeekOptions.one_based_numbering.
```
```suggestion
/// depending on DayOfWeekOptions.count_from_zero.
```

jorisvandenbossche replied:
BTW, I find the "Week numbering can start with 0 or 1" a bit confusing, as you don't really start to count from 0 (it's not that this represents a "full" week, and that the days at the end of the year (a potential week 53) are also coded as week 0).

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -1006,6 +1080,15 @@ const FunctionDoc iso_week_doc{
+    "Extract week of year number",
+    ("First week has the majority (4 or more) of its days in January.\n"
+     "Year can have 52 or 53 weeks. Week numbering can start with 0 or 1 using "
+     "DayOfWeekOptions.one_based_numbering.\n"
```
```suggestion
     "DayOfWeekOptions.count_from_zero.\n"
```

jorisvandenbossche commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -1006,6 +1080,15 @@ const FunctionDoc iso_week_doc{
+    ("First week has the majority (4 or more) of its days in January.\n"
+     "Year can have 52 or 53 weeks. Week numbering can start with 0 or 1 using "
+     "DayOfWeekOptions.one_based_numbering.\n"
+     "Returns an error if timestamp has a defined timezone. Null values return null."),
```
Why does not work for timezone data? (or is this a copy-paste left-over?)

rok replied:
Leftover from before this would generally be possible. Removed.

rok commented:
@jorisvandenbossche thanks for the review!

I think there are two more open questions here:
1. Do we fully implement MySQL functionality (8 different modes for computing week) in full or not? Currently this proposal does. [See discussion.](https://github.com/apache/arrow/pull/11026#pullrequestreview-747344728) 
2. We can remove one boolean check (per value) from the Week kernel by templating. Should we?

lidavidm reviewed (approved):
I'm OK with the implementation (though it would be nice to consolidate the iso_week kernel into this); I would lean towards offering all the possible options but I suppose most systems only offer iso_week and maybe us_week.

pitrou commented:
I think we can offer two compute functions "iso_week" and "us_week" that simply redirect to this kernel with appropriate settings. That would make those convenience functions automatically visible from bindings such as Python.

pitrou commented on python/pyarrow/_compute.pyx:
```diff
@@ -1091,6 +1091,22 @@ class AssumeTimezoneOptions(_AssumeTimezoneOptions):
+
+
+class WeekOptions(_WeekOptions):
+    def __init__(self, week_starts_monday=True, count_from_zero=False,
```
These should be keyword-only arguments, as `WeekOptions(False, False, True)` is basically write-only code.

rok replied:
Done.

rok commented:
@lidavidm @pitrou I've removed the `ISOWeek` kernel and added `iso_week` and `us_week` functions that use `week` with specific options.

rok commented:
R [CI is reporting](https://github.com/apache/arrow/pull/11026/checks?check_run_id=3687105253):
```
── Error (test-dplyr-lubridate.R:80:3): extract isoweek from timestamp ─────────
Error: Invalid: Attempted to initialize KernelState from null FunctionOptions
Backtrace:
     █
  1. ├─arrow:::expect_dplyr_equal(...) test-dplyr-lubridate.R:80:2
  2. │ ├─testthat::expect_warning(...) helper-expectation.R:101:4
  3. │ │ └─testthat:::quasi_capture(enquo(object), label, capture_warnings)
  4. │ │   ├─testthat:::.capture(...)
  5. │ │   │ └─base::withCallingHandlers(...)
  6. │ │   └─rlang::eval_bare(quo_get_expr(.quo), quo_get_env(.quo))
  7. │ └─rlang::eval_tidy(expr, rlang::new_data_mask(rlang::env(input = record_batch(tbl))))
  8. ├─input %>% mutate(x = isoweek(datetime)) %>% collect()
  9. ├─dplyr::collect(.)
 10. └─arrow:::collect.arrow_dplyr_query(.)
 11.   └─arrow:::do_exec_plan(x)
 12.     └─plan$Build(.data)
 13.       └─node$Project(projection)
 14.         ├─self$preserve_sort(ExecNode_Project(self, cols, names(cols)))
 15.         └─arrow:::ExecNode_Project(self, cols, names(cols))
── Error (test-dplyr-lubridate.R:205:3): extract isoweek from date ─────────────
Error: Invalid: Attempted to initialize KernelState from null FunctionOptions
```

I suppose we then need ISOWeek and USWeek declared with options in the API.

lidavidm commented:
Yeah, we'd need a case here: https://github.com/apache/arrow/blob/075c6c6c0081688f10e101bb781b81d0e31a8986/r/src/compute.cpp#L129

lidavidm commented on r/src/compute.cpp:
```diff
@@ -318,14 +318,42 @@ std::shared_ptr<arrow::compute::FunctionOptions> make_compute_options(
 
+  if (func_name == "iso_week") {
+    using Options = arrow::compute::WeekOptions;
+    return std::make_shared<Options>(true, false, false);
```
nit: place `/*count_from_zero=*/`, etc. comments here to make it clear?

rok replied:
Switched to `::ISODefaults()` so no longer necessary.

lidavidm commented on docs/source/cpp/compute.rst:
```diff
@@ -1356,11 +1360,19 @@ For timestamps inputs with non-empty timezone, localized timestamp components wi
+  starts with the first ISO week. ISO week starts on Monday.
   See `ISO 8601 week date definition`_ for more details.
-* \(3) Output is a ``{"iso_year": output type, "iso_week": output type, "iso_day_of_week":  output type}`` Struct.
+* \(3) First US week has the majority (4 or more) of it's days in January. US year
+  starts with the first US week. US week starts on Sunday.
+* \(4) Returns week number allowing for setting several parameters.
+  :member:`WeekOptions::week_starts_monday` sets what day does the week start with (Monday=true, Sunday=false).
+  :member:`WeekOptions::count_from_zero` sets that dates from current year that fall into last ISO week of
+  the previous year return 0 if true and 52 or 53 if false.
+  :member:`WeekOptions::first_week_is_fully_in_year` sets if the first week is fully in January (true), or is a
+  week that begins on December 29, 30, or 31 considered to be the first week of the new year (false)?
+* \(5) Output is a ``{"iso_year": output type, "iso_week": output type, "iso_day_of_week":  output type}`` Struct.
```
```suggestion
* \(3) First US week has the majority (4 or more) of its days in January. US year
  starts with the first US week. US week starts on Sunday.
* \(4) Returns week number allowing for setting several parameters.
  If :member:`WeekOptions::week_starts_monday` is true, the week starts with Monday, else Sunday if false.
  If :member:`WeekOptions::count_from_zero` is true, dates from the current year that fall into the last ISO week
  of the previous year are numbered as week 0, else week 52 or 53 if false.
  If :member:`WeekOptions::first_week_is_fully_in_year` is true, the first week (week 1) must fully be in January;
  else if false, a week that begins on December 29, 30, or 31 is considered the first week of the new year.
* \(5) Output is a ``{"iso_year": output type, "iso_week": output type, "iso_day_of_week":  output type}`` Struct.
```

lidavidm replied:
Some wording tweaks just to make this flow a little better to me (feel free to accept/reject/reword as needed).

lidavidm commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -1135,10 +1210,24 @@ void RegisterScalarTemporal(FunctionRegistry* registry) {
 
-  auto iso_week = MakeTemporal<ISOWeek, TemporalComponentExtract, Int64Type>(
-      "iso_week", {WithDates, WithTimestamps}, int64(), &iso_week_doc);
+  static const auto default_iso_week_options = WeekOptions(true, false, false);
```
Similarly here, can we put comments for the parameter literals? Or actually, it might be nice to define WeekOptions::IsoDefaults and WeekOptions::UsDefaults which could then be used here and in the R bindings.

rok replied:
Added `WeekOptions::ISODefaults` and `WeekOptions::USDefaults`.

lidavidm commented on cpp/src/arrow/compute/api_scalar.h:
```diff
@@ -1006,17 +1023,20 @@ ARROW_EXPORT Result<Datum> DayOfYear(const Datum& values, ExecContext* ctx = NUL
 /// \note API not yet finalized
-ARROW_EXPORT Result<Datum> ISOWeek(const Datum& values, ExecContext* ctx = NULLPTR);
+ARROW_EXPORT Result<Datum> Week(const Datum& values, WeekOptions options = WeekOptions(),
+                                ExecContext* ctx = NULLPTR);
```
Also add a helper for USWeek?

rok replied:
Done.

lidavidm replied:
It appears to be missing its counterpart in api_scalar.cc.

rok replied:
Added.

rok commented:
> Yeah, we'd need a case here:
> 
> https://github.com/apache/arrow/blob/075c6c6c0081688f10e101bb781b81d0e31a8986/r/src/compute.cpp#L129

@lidavidm oh, didn't realize it had to be there.
Maybe would be good to add R tests as well then?

lidavidm commented on cpp/src/arrow/compute/kernels/scalar_temporal.cc:
```diff
@@ -364,29 +379,68 @@ struct ISOYear {
 // ----------------------------------------------------------------------
 // Extract ISO week from temporal types
```
These comments might need updating (at least 'Extract week from temporal types', the definition of ISO week can be left here)

rok replied:
I've just removed "ISO". Do you thing a bigger change is needed?

lidavidm replied:
This is good, thanks.

lidavidm reviewed (commented):
Some nits on style/wording.

If you're up to add R tests, I think that would be appreciated, but otherwise we can put that in a follow-up JIRA (since there might be other work to bind the new kernels to dplyr features or whatnot).

lidavidm reviewed (approved):
Thanks, LGTM.

rok commented:
Thanks for the review @lidavidm ! :)

pitrou reviewed (approved):
Rebased, will merge

rok commented:
Thanks @pitrou @ianmcook @jorisvandenbossche @lidavidm !
